torchrl/objectives/value/functional.py

```diff
@@ -712,26 +832,26 @@ def td_lambda_return_estimate(
         raise RuntimeError(
             "rolling_gamma=False is expected only with time-sensitive gamma or lambda values"
         )
-
     if rolling_gamma:
-        gamma = gamma * not_done
         g = next_state_value[..., -1, :]
         for i in reversed(range(T)):
+            dn = done[..., i, :].int()
+            nv = next_state_value[..., i, :]
+            lmd = lmbda[..., i, :]
+            g = g * (1 - dn) + nv * dn
             g = returns[..., i, :] = reward[..., i, :] + gamma[..., i, :] * (
-                (1 - lmbda[..., i, :]) * next_state_value[..., i, :]
-                + lmbda[..., i, :] * g
+                (1 - lmd) * nv + lmd * g
             )
     else:
         for k in range(T):
             g = next_state_value[..., -1, :]
             _gamma = gamma[..., k, :]
             _lambda = lmbda[..., k, :]
-            nd = not_done
-            _gamma = _gamma.unsqueeze(-2) * nd
             for i in reversed(range(k, T)):
-                g = reward[..., i, :] + _gamma[..., i, :] * (
-                    (1 - _lambda) * next_state_value[..., i, :] + _lambda * g
-                )
+                dn = done[..., i, :].int()
+                nv = next_state_value[..., i, :]
+                g = g * (1 - dn) + nv * dn
+                g = reward[..., i, :] + _gamma * ((1 - _lambda) * nv + _lambda * g)
```

Comment:
it seems like here the logic changed quite a bit.
I suggest writing some in-line comments for this very crucial bits

Reply:
Only in apparence, before we were integrating the not_done in the gamma, now we do it on the gain.
The reason why this is needed is that we want for the next value to be equal to next_state_value when done=True. 
Before this was working because the next value was going to be always 0 when done=True as truncated was also True, but this does not hold anymore.

Happy to give a small comment on what we're doing oc!